worker: add eventLoopUtilization()

trevnorris opened the pull request:
NOTE: While this is working as it should, I feel the implementation may be a bit clunky. Specifically the one TODO about thread-safe access to a performance mark. Looking for feedback on how this could be improved.

-----

worker: add eventLoopUtilization()

Allow calling eventLoopUtilization() directly on a worker thread:

```js
const worker = new Worker('./foo.js');
const elu = worker.performance.eventLoopUtilization();
setTimeout(() => {
  worker.performance.eventLoopUtilization(elu);
}, 10);
```

Add a new performance object on the Worker instance that will hopefully
one day hold all the other performance metrics, such as nodeTiming.

Include benchmarks and tests.

Also include a commit that improves tests for `perf_hooks.performance.eventLoopUtilization()`.

-----

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

juanarbol reviewed (approved):
LGTM

jasnell commented on src/node_worker.cc:
```diff
@@ -741,6 +741,39 @@ void Worker::TakeHeapSnapshot(const FunctionCallbackInfo<Value>& args) {
+    return args.GetReturnValue().Set(-1);
+
+  // TODO(trevnorris): This isn't technically thread safe.
+  double loop_start_time = w->env_->performance_state()->milestones[
```
The key question here then would be if the loop start milestone is guaranteed to be set before user code is permitted to call LoopStartTime. I suspect that it's not. Curious what @addaleax thinks here.

addaleax replied:
> The reason this (probably) works is because it's only doing a read, and the value is only written to once. So if the state hasn't synchronized it'll return `{active: 0, idle: 0, utilization: 0}`.

It’s not necessarily that easy – we should not assume that 64-bit reads & writes are atomic on every platform, I think.

The easiest way to work around this would probably be to only allow calling this function if the `'online'` event has been received on the Worker instance, and if not, skipping this call into C++. We could then assert here that `loop_start_time > 0`.

trevnorris replied:
@addaleax 64-bit reads/writes are always atomic on 64-bit systems if the memory is aligned. So, since we support 32-bit Windows builds, I guess this isn't solved so easily. :-(

addaleax replied:
@trevnorris That’s true for x64, but we’re also supporting arm and ppc :)

jasnell commented:
Looks good but let's work through the possible threadsafety issue on the performance mark before sign off

juanarbol commented:
I'm agree with @jasnell, sorry about the rush, haha

addaleax commented on src/node_worker.cc:
```diff
@@ -741,6 +741,39 @@ void Worker::TakeHeapSnapshot(const FunctionCallbackInfo<Value>& args) {
+  // Using w->is_stopped() here leads to a deadlock, and checking is_stopped()
+  // before locking the mutex is a race condition. So manually do the same
+  // check.
+  if (w->stopped_ || w->env_ == nullptr || w->env_->is_stopping())
```
The last check here would be equivalent to the first one if `w->env_ != nullptr`

addaleax commented on lib/internal/worker.js:
```diff
@@ -415,6 +423,37 @@ function makeResourceLimits(float64arr) {
+    return { idle, active, utilization: active / (idle + active) };
+  }
+
+  const idle = this[kHandle].loopIdleTime();
```
Keep in mind that `idle` can be `-1` even if `this[kLoopStartTime]` was positive.

trevnorris replied:
Whoops. Thanks.

trevnorris commented:
@addaleax Thanks for the review. I've made fixes for everything, including only allowing to call `LoopStartTime` after the `'online'` event.

trevnorris commented:
@addaleax Fix the benchmark issue. Which made me realize that because reading the ELU of a worker can't be done until after the `'online'` event, it now required the event loop to iterate at least once before it can be read.

This isn't optimal, but I can't think of a better solution for now. In the future I wanted to make reading/writing to all the milestones atomic so we could add `worker.performance.nodeTiming` as well. If that ever happens, we can remove the current limitation.

addaleax commented:
>  Which made me realize that because reading the ELU of a worker can't be done until after the `'online'` event, it now required the event loop to iterate at least once before it can be read.

I think the current approach of returning an object with `0` values is just fine.

> In the future I wanted to make reading/writing to all the milestones atomic so we could add `worker.performance.nodeTiming` as well.

Keep in mind that even if reading a single value might be atomic, reading the entire group of timing values would not be.

addaleax commented:
@nodejs/worker

benjamingr commented on doc/api/worker_threads.md:
```diff
@@ -913,6 +913,26 @@ Calling `unref()` on a worker will allow the thread to exit if this is the only
+values for the worker instance.
+
+Currently the event loop utilization of a worker instance can only be retrieved
+after the [`'online'` event][] event has fired.
```
Add note about what happens if you call it before `online` has fired?

Trott replied:
```suggestion
after the [`'online'` event][] event is emitted.
```

benjamingr commented on lib/internal/worker.js:
```diff
@@ -223,6 +227,12 @@ class Worker extends EventEmitter {
       hasStdin: !!options.stdin
     }, transferList);
+    // Use this to cache the Worker's loopStart value once available.
+    this[kLoopStartTime] = -1;
```
Put the -1 in a const?

trevnorris replied:
@benjamingr Sorry, I don't follow what you mean.

benjamingr commented on lib/internal/worker.js:
```diff
@@ -415,6 +426,50 @@ function makeResourceLimits(float64arr) {
 }
 
+function eventLoopUtilization(util1, util2) {
+  // TODO(trevnorris): Works to solve the thread-safe read/write issue of
```
Would it make sense for the API instead to be emitted on `online` so:

 - The worker emits `online` with an "stats" argument
 - The stats argument has this `eventLoopUtilization` method the user can call

That would make it impossible to use this API incorrectly and remove a bit of the weirdness.

addaleax replied:
That seems like it makes accessing the API more difficult, with little benefit…

benjamingr replied:
@addaleax
Well, the benefit is that it would be impossible to misuse the API. Currently this API has a (relatively small) footgun if you call `eventLoopUtilization` before `online` is fired. That eliminates the footgun.

Note that the API isn't really that bad:

```js
const worker = new Worker(...);
const controller = await once(worker, 'online');
console.log(controller.eventLoopUtilization()); // same thing as currently on worker now
```

addaleax replied:
@benjamingr I think my disagreement is more about whether this is a footgun or not :) If you call it before the `'online'` event, you don’t get inaccurate data, you get no ELU data – which is *accurate*, because (up to race conditions) there actually is no ELU data available.

benjamingr replied:
If you think the distinction between "no data" and "0 data" here isn't important - I will concede.

trevnorris replied:
@benjamingr I plan on figuring out a way to remove the limitation of not being able to query the ELU until the `'online'` event has been emitted.

benjamingr commented on lib/internal/worker.js:
```diff
@@ -415,6 +426,50 @@ function makeResourceLimits(float64arr) {
+  // has had a chance to turn. So it will be impossible to read the ELU of
+  // a worker thread immediately after it's been created.
+  if (!this[kIsOnline]) {
+    return { idle: 0, active: 0, utilization: 0 };
```
If we keep this API, I really prefer it if we didn't return `0`s to indicate "not yet online" here and instead throw an error. I can see users running into footguns here.

addaleax replied:
This currently agrees with the behavior on the main thread. I think throwing an error would be the (way) more unexpected result.

benjamingr commented on lib/internal/worker.js:
```diff
@@ -415,6 +426,50 @@ function makeResourceLimits(float64arr) {
+  }
+
+  // Cache loopStart, since it's only written to once.
+  if (this[kLoopStartTime] <= 0) {
```
=== -1 (rather than <=0) ? (also, const?)

benjamingr commented:
Also, I did not actually check out the branch and play with it so I didn't LGTM as this already has enough approvals - but if that would help/is needed here I am happy to do that.

To be clear (eventhough thankfully our process doesn't require this anymore): none of my comments are blocking and I am +1 on these changes in general.

Trott commented on doc/api/worker_threads.md:
```diff
@@ -913,6 +913,26 @@ Calling `unref()` on a worker will allow the thread to exit if this is the only
+The same call as [`perf_hooks` `eventLoopUtilization()`][], except returns the
+values for the worker instance.
+
+Currently the event loop utilization of a worker instance can only be retrieved
```
```suggestion
The event loop utilization of a worker instance can only be retrieved
```

Trott commented on doc/api/worker_threads.md:
```diff
@@ -940,6 +960,8 @@ active handle in the event system. If the worker is already `unref()`ed calling
 [`fs.close()`]: fs.md#fs_fs_close_fd_callback
 [`fs.open()`]: fs.md#fs_fs_open_path_flags_mode_callback
 [`markAsUntransferable()`]: #worker_threads_worker_markasuntransferable_object
+[`'online'` event]: #worker_threads_event_online
```
I don't think this one is correctly ordered (because `'` should put it lexically before any letter) but I'm also surprised the linter isn't complaining about it. 
¯\\_(ツ)_/¯

trevnorris replied:
Thanks. I'll fix it anyways.

Trott commented on doc/api/worker_threads.md:
```diff
@@ -913,6 +913,26 @@ Calling `unref()` on a worker will allow the thread to exit if this is the only
 active handle in the event system. If the worker is already `unref()`ed calling
 `unref()` again will have no effect.
 
+### `worker.performance.eventLoopUtilization([utilization1[, utilization2]])`
```
Entries are in alphabetical order. So I want to suggest moving this below `worker.getHeapSnapshot()` and above `worker.postMessage()`, but....

Do we need to add an entry (even if it's just a single sentence) for `worker.performance` and make this a subheader of that entry? If so, where would that go in correct order? @nodejs/documentation

trevnorris replied:
@Trott Eventually I'd like the `worker.performance` object to contain the `nodeTiming` object. If we don't think that will ever happen then I'd suggest we change the API to `worker.eventLoopUtilization()`.

Trott reviewed (commented):
My comments on the docs are not blocking. Those things can be addressed in subsequent commits after landing if need be.

trevnorris commented:
@addaleax 
> Keep in mind that even if reading a single value might be atomic, reading the entire group of timing values would not be.

My other thought was b/c all the `nodeTiming` values are only ever written to once that it should be somehow possible to propagate the value to the thread that spawned the worker, but I honestly can't think of how to do this.

trevnorris commented:
@Trott Fixed the documentation position gave it a bullet list and more examples.

@addaleax Found and fixed a bug thanks to something you pointed out. Updated the tests with it and cleaned up the PR.

Think this is ready.

trevnorris commented:
@jasnell Fixed the lint issue. Sorry about that.

trevnorris commented:
@jasnell It looks like alpine isn't pulling the latest code? This failure:
```
not ok 2629 parallel/test-worker-eventlooputil
  ---
    node:assert:388
        throw err;
        ^
    AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:
    
      assert.ok(idleActive(wElu) >= 50)
    
        at Timeout.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/alpine-last-latest-x64/test/parallel/test-worker-eventlooputil.js:79:12)
```
That code doesn't exist anymore. I changed it to `assert.ok(idleActive(wElu) >= 45)` to get around the timing problem.

Trott commented:
It looks like maybe this breaks `tools/test.py --worker test/parallel/test-bootstrap-modules`? (Basing this on CI results only.)

trevnorris commented:
@Trott Since I'm only using `performance.now()` would the better solution be to just copy out the `function now()` code (since it's only a few lines of code), or change the test so it expects those modules?

trevnorris commented:
@Trott In order to reduce the Worker footprint as much as possible, I opt'd to copy out the code for `performance.now()`. That still required me to get `timeOrigin`, but better than nothing. Please give CI one more run.

Flarna commented:
@trevnorris Last CI run shows a fail which is maybe related https://ci.nodejs.org/job/node-test-commit-osx/36889/nodes=osx1015/testReport/junit/(root)/test/parallel_test_performance_eventlooputil/

trevnorris commented:
@Flarna That failure shouldn't be possible in any scenario. It basically evaluates to:
```js
const t = performance.now();
while (performance.now() - t < 50);
assert.ok(performance.now() - t > 0);
```
The only difference is that it uses `nodeTiming.loopStart` as an offset of when the event loop was entered. But since I force the loop to have started and accumulated `idle` time before the test runs it would imply that `loopStart` has a bug.

I've made another alteration to the tests, and made sure the failure string outputs the values so I can see how it's failing, if it fails. I also fixed the code coverage regression. Please git it another go in CI.

trevnorris commented:
@Flarna Is there anything else I need to do?

Flarna commented:
@trevnorris I don't think so. Landing via bot failed so it needs to be done the old way. I will be back on my working machine tomorrow.

Flarna commented:
Landed in 04d16646a089

trevnorris commented:
Awesome. Thank you much @Flarna.
